src/server/worker/pipeline/transform/worker.go

```diff
@@ -347,6 +304,125 @@ func handleDatumSet(driver driver.Driver, logger logs.TaggedLogger, task *DatumS
 	})
 }
 
+func handleDatumSetBatching(ctx context.Context, driver driver.Driver, logger logs.TaggedLogger, task *DatumSetTask, status *Status, cacheClient *pfssync.CacheClient, di datum.Iterator, setOpts []datum.SetOption) error {
+	nextChan, setupChan := make(chan error), make(chan []string)
+	status.nextChan, status.setupChan = nextChan, setupChan
+	// Set up the restart mechanism for the user code.
+	var cancel context.CancelFunc
+	var errChan chan error
+	stop := func() {
+		if cancel != nil {
+			cancel()
+			<-errChan
+		}
+	}
+	defer stop()
+	start := func() error {
+		var cancelCtx context.Context
+		cancelCtx, cancel = context.WithCancel(ctx)
+		errChan = make(chan error, 1)
+		go func() {
+			err := driver.RunUserCode(cancelCtx, logger, nil)
+			if err == nil {
+				err = errors.New("user code exited prematurely")
+			}
+			errChan <- err
+			close(errChan)
+		}()
+		select {
+		case <-nextChan:
+			return nil
+		case <-ctx.Done():
+			return errors.EnsureStack(ctx.Err())
+		}
+	}
+	// Start the user code, then iterate through the datums.
+	if err := start(); err != nil {
+		return errors.Wrap(err, "error starting user code")
+	}
+	return forEachDatum(ctx, driver, logger, task, status, cacheClient, di, setOpts, func(ctx context.Context, logger logs.TaggedLogger, env []string) (retErr error) {
+		defer func() {
+			// Restart the user code if an error occurred.
+			if retErr != nil {
+				stop()
+				if err := start(); err != nil {
+					retErr = multierror.Append(retErr, errors.Wrap(err, "error restarting user code"))
+				}
+			}
+		}()
+		select {
+		case status.setupChan <- env:
+		case err := <-errChan:
+			return errors.Wrap(err, "error running user code")
+		case <-ctx.Done():
+			return errors.EnsureStack(ctx.Err())
+		}
+		select {
+		case err := <-nextChan:
+			return err
+		case err := <-errChan:
+			return errors.Wrap(err, "error running user code")
+		case <-ctx.Done():
+			return errors.EnsureStack(ctx.Err())
+		}
+	})
+}
+
+type datumCallback = func(ctx context.Context, logger logs.TaggedLogger, env []string) error
+
+// TODO: There are way too many parameters here. Consider grouping them into a reasonable struct or storing some of these in the driver.
+func forEachDatum(ctx context.Context, driver driver.Driver, logger logs.TaggedLogger, task *DatumSetTask, status *Status, cacheClient *pfssync.CacheClient, di datum.Iterator, setOpts []datum.SetOption, cb datumCallback) error {
+	// TODO: Can this just be refactored into the datum package such that we don't need to specify a storage root for the sets?
+	// The sets would just create a temporary directory under /tmp.
```

Comment:
I don't think we should be building things that assume they can write to a path without being passed that path as a parameter.  It makes it difficult to test multiple copies of the thing concurrently, and it makes the component more difficult to reason about because it has secret global state from the filesystem.

Reply:
This was just a copied over TODO. Generally making the root directory / prefix configurable makes sense.